react-native-hub-app/src/checklist.js

```diff
@@ -0,0 +1,449 @@
+/**
+ * Sample React Native App
+ * https://github.com/facebook/react-native
+ *
+ * @format
+ * @flow strict-local
+ */
+
+import React from 'react';
+import {
+  FlatList,
+  StyleSheet,
+  View,
+  Text,
+  TouchableOpacity,
+  AsyncStorage,
+} from 'react-native';
+import {Client, Where} from '@textile/threads-client';
+import {ThreadID} from '@textile/threads-id';
+import {Buckets, Context, createAPISig} from '@textile/textile';
+import {Libp2pCryptoIdentity} from '@textile/threads-core';
+import {USER_API_SECRET, USER_API_KEY, API_URL} from 'react-native-dotenv';
+
+const MAX_STEPS = 2;
+const IDENTITY_KEY = 'identity';
+const CONTEXT_KEY = 'context';
+const TOKEN_KEY = 'token';
+const ASTRONAUT_THREAD_ID = 'astronaut_id';
+const sleep = (m) => new Promise((r) => setTimeout(r, m));
+
+class CheckList extends React.Component {
+  constructor(props) {
+    super(props);
+  }
+
+  // you could also do this, so no constructor needed
+  state = {
+    steps: [
+      {key: 'Step 0', name: 'Prepare Identity & API Token', status: 0},
+      {key: 'Step 1', name: 'Setup ThreadDB', status: 0},
+      {key: 'Step 2', name: 'Add Instance to Collection', status: 0},
+      {key: 'Step 3', name: 'Query from our Collection', status: 0},
+      {key: 'Step 4', name: 'List existing Buckets', status: 0},
+    ],
+    step: 0,
+    errorMessage: '',
+  };
+
+  incrementStatus(index) {
+    const steps = this.state.steps;
+    const data = steps[index];
+    if (data.status >= MAX_STEPS) {
+      return;
+    }
+    data.status = data.status += 1;
+    steps[index] = data;
+    this.setState({
+      steps: steps,
+    });
+  }
+
+  async generateIdentity() {
+    let idStr = await AsyncStorage.getItem(IDENTITY_KEY);
+    if (idStr) {
+      return await Libp2pCryptoIdentity.fromString(idStr);
+    } else {
+      const id = await Libp2pCryptoIdentity.fromRandom();
+      idStr = id.toString();
+      await AsyncStorage.setItem(IDENTITY_KEY, idStr);
+      return id;
+    }
+  }
+
+  async getUserToken(database, id) {
+    let token = await AsyncStorage.getItem(TOKEN_KEY);
+    if (token) {
+      return token;
+    }
+    token = await database.getToken(id);
+    await AsyncStorage.setItem(TOKEN_KEY, token);
+    return token;
+  }
+
+  async getContext() {
+    // Pull the stored context to reuse if available && valid date
+    let contextStr = await AsyncStorage.getItem(CONTEXT_KEY);
+    if (contextStr) {
+      const ctxJson = JSON.parse(contextStr);
+      if (
+        ctxJson['x-textile-api-sig-msg'] && (Date.parse(ctxJson['x-textile-api-sig-msg'])) > (new Date()).getTime()) {
+        // Not expired
+        const ctx = Context.fromJSON(ctxJson, API_URL);
+        return ctx;
+      }
+    }
+  }
+
+  async getAstronautThread(db) {
+    let idStr;// = await AsyncStorage.getItem(ASTRONAUT_THREAD_ID);
```

Comment:
If I uncomment this line and allow my Thread ID to be restored from string, I start to get errors from DB. For example, step 3 will next try to insert an instance. If I do this on an existing threadId, it will always give me the result `Failure: Type not convertable to Uint8Array'

Reply:
Hmmm, that sounds like an issue creating the protobuf. Possible to check the thread if string? Maybe it’s not being serialized properly?

Reply:
ah It could also be that thread Id is undefined when calling NewDB?

Reply:
threadid exists when from random and when restored. when restored, i can do a `JSON.stringify(orig) === JSON.stringify(restored)` and it's true.

Reply:
db.newDB(restored) === `Failure: Type not convertable to Uint8Array`

Reply:
what it is attempting here is to rejoin a db that the identity already owns. if i don't call `newDB` and just try to run an insert against the threaddb, i get the same error.

Reply:
converting this ship to typescript... so much for vanilla js example right  now

Reply:
okay, no red flags in typescript